README
Stamps.com API for Python


The Idea
========

The goal of this API is to provide a thin Python wrapper around the Stamps.com
Web Services API. As to why anyone would be generating big W-S APIs in the 21st
century? Don't get me started.

This API gives you a flexible configuration mechanism, calls for the more
common operations, cached authenticator token, conversation out-of-sync
detection, and examples as test cases.

This API depends on Python "Suds" and exerts minimal opinion on the format or
results of the Stamps.com operations. In other words, the naming conventions
for Suds-based I/O are not Pythonic and are dictated, instead, by the backing
WSDL. You'll probably want to extend the API here with your own so you can bury
all of the weird syntax under your own version of awesomeness.


The Installation
================

Simple with pip:

$ pip install stamps.py

Easy with setuptools (but you really shouldn't):

$ easy_install stamps.py


The Interface
=============

These and other public functions are documented in
stamps.services.StampsService:

add_postage
    Add postage to your account so you can start creating labels.

create_*
    Convenience methods for creating various call objects. In general, these
    will be bare objects that need to be populated with detail information
    based on the call you wish to make. See the get_* functions in tests.py for
    details.

get_account
    Account information accessor.

get_label
    Get a shipping label from one address to another based on a given rate.

get_rates
    Get the available rates for a given shipping object.

get_tracking
    Get the tracking events for a given label ID.

register_account
    Register a new account with Stamps.com.

remove_label
    Cancel a shipping label (triggers a postage refund with Stamps.com).


The Instantiation
=================

The best way to get started is to digest the code in tests.py, which uses a
file to initialize a test StampsConfiguration. Below is a simple example that
uses the API to retrieve account information. Register with
developer.stamps.com to get your own integration ID, username, and password.

    from stamps.config import StampsConfiguration
    from stamps.services import StampsService

    integration_id = "XXXXXXXX-1111-2222-3333-YYYYYYYYYYYY"
    username = "stampy"
    password = "secret"
    configuration = StampsConfiguration(integration_id=integration_id,
        username=username, password=password)
    service = StampsService(configuration=configuration)
    account = service.get_account()


CHANGES
Changelog
---------

Version 0.9.1
^^^^^^^^^^^^^

* Validate `get_rates` service response.

Version 0.9
^^^^^^^^^^^

* Add customs support for international shipping.

Version 0.8
^^^^^^^^^^^

* Updated to Stamps.com v49 web services.
* Required suds-jurko.

Version 0.7.2
~~~~~~~~~~~~~

* Updated to Stamps.com v36 web services.

Version 0.7.1
^^^^^^^^^^^^^

* Updated to Stamps.com v35 web services.

Version 0.7
^^^^^^^^^^^

* Updated to Stamps.com v34 web services.
* Mapped xs:decimal to custom XDecimal type.

Version 0.6
^^^^^^^^^^^

* Updated to Stamps.com v33 web services.
* Support cross-platform WSDL file URL.

Version 0.5
^^^^^^^^^^^

* Updated to Stamps.com v31 web services.

Version 0.4
^^^^^^^^^^^

* Updated to Stamps.com v29 web services.
* Added quick start example to the README.

Version 0.3
^^^^^^^^^^^

* Enhanced service configuration.
* Added the README.
* Updated to Stamps.com v26 web services.

Version 0.2
^^^^^^^^^^^

* Updated to Stamps.com v25 web services.

Version 0.1
^^^^^^^^^^^

* Initial public release.


stamps/tests.py
# -*- coding: utf-8 -*-
"""
    stamps.tests
    ~~~~~~~~~~~~

    Stamps.com API tests.

    :copyright: 2014 by Jonathan Zempel.
    :license: BSD, see LICENSE for more details.
"""

from .config import StampsConfiguration
from .services import AuthenticatorPlugin, StampsService
from datetime import date, datetime
from time import sleep
from unittest import TestCase
import logging
import os


logging.basicConfig()
logging.getLogger("suds.client").setLevel(logging.DEBUG)
file_path = os.path.abspath(__file__)
directory_path = os.path.dirname(file_path)
file_name = os.path.join(directory_path, "tests.cfg")
CONFIGURATION = StampsConfiguration(wsdl="testing", file_name=file_name)


def get_rate(service):
    """Get a test rate.

    :param service: Instance of the stamps service.
    """
    ret_val = service.create_shipping()
    ret_val.ShipDate = date.today().isoformat()
    ret_val.FromZIPCode = "94107"
    ret_val.ToZIPCode = "20500"
    ret_val.PackageType = "Package"
    rate = service.get_rates(ret_val)[0]
    ret_val.Amount = rate.Amount
    ret_val.ServiceType = rate.ServiceType
    ret_val.DeliverDays = rate.DeliverDays
    ret_val.DimWeighting = rate.DimWeighting
    ret_val.Zone = rate.Zone
    ret_val.RateCategory = rate.RateCategory
    ret_val.ToState = rate.ToState
    add_on = service.create_add_on()
    add_on.AddOnType = "US-A-DC"
    ret_val.AddOns.AddOnV6.append(add_on)

    return ret_val


def get_from_address(service):
    """Get a test 'from' address.

    :param service: Instance of the stamps service.
    """
    address = service.create_address()
    address.FullName = "Pickwick & Weller"
    address.Address1 = "300 Brannan St."
    address.Address2 = "Suite 405"
    address.City = "San Francisco"
    address.State = "CA"

    return service.get_address(address).Address


def get_to_address(service):
    """Get a test 'to' address.

    :param service: Instance of the stamps service.
    """
    address = service.create_address()
    address.FullName = "POTUS"
    address.Address1 = "1600 Pennsylvania Avenue NW"
    address.City = "Washington"
    address.State = "DC"

    return service.get_address(address).Address


class StampsTestCase(TestCase):

    initialized = False

    def setUp(self):
        if not StampsTestCase.initialized:
            self.service = StampsService(CONFIGURATION)
            StampsTestCase.initalized = True

    def _test_0(self):
        """Test account registration.
        """
        registration = self.service.create_registration()
        type = self.service.create("CodewordType")
        registration.Codeword1Type = type.Last4SocialSecurityNumber
        registration.Codeword1 = 1234
        registration.Codeword2Type = type.Last4DriversLicense
        registration.Codeword2 = 1234
        registration.PhysicalAddress = get_from_address(self.service)
        registration.MachineInfo.IPAddress = "127.0.0.1"
        registration.Email = "sws-support@stamps.com"
        type = self.service.create("AccountType")
        registration.AccountType = type.OfficeBasedBusiness
        result = self.service.register_account(registration)
        print result

    def _test_1(self):
        """Test postage purchase.
        """
        transaction_id = datetime.now().isoformat()
        result = self.service.add_postage(10, transaction_id=transaction_id)
        transaction_id = result.TransactionID
        status = self.service.create_purchase_status()
        seconds = 4

        while result.PurchaseStatus in (status.Pending, status.Processing):
            seconds = 32 if seconds * 2 >= 32 else seconds * 2
            print "Waiting {0:d} seconds to get status...".format(seconds)
            sleep(seconds)
            result = self.service.get_postage_status(transaction_id)

        print result

    def test_2(self):
        """Test label generation.
        """
        self.service = StampsService(CONFIGURATION)
        rate = get_rate(self.service)
        from_address = get_from_address(self.service)
        to_address = get_to_address(self.service)
        transaction_id = datetime.now().isoformat()
        label = self.service.get_label(from_address, to_address, rate,
                transaction_id=transaction_id)
        self.service.get_tracking(label.StampsTxID)
        self.service.get_tracking(label.TrackingNumber)
        self.service.remove_label(label.StampsTxID)
        print label

    def test_3(self):
        """Test authentication retry.
        """
        self.service.get_account()
        authenticator = self.service.plugin.authenticator
        self.service.get_account()
        self.service.plugin.authenticator = authenticator
        result = self.service.get_account()
        print result

    def test_4(self):
        """Test custom authenticator plugin.
        """
        class InvalidAuthenticatorPlugin(object):
            pass

        with self.assertRaises(TypeError):
            StampsService(CONFIGURATION, InvalidAuthenticatorPlugin)

        class ValidAuthenticatorPlugin(AuthenticatorPlugin):
            pass

        StampsService(CONFIGURATION, ValidAuthenticatorPlugin)


stamps/services.py
# -*- coding: utf-8 -*-
"""
    stamps.services
    ~~~~~~~~~~~~~~~

    Stamps.com services.

    :copyright: 2014 by Jonathan Zempel.
    :license: BSD, see LICENSE for more details.
"""

from decimal import Decimal
from logging import getLogger
from re import compile
from suds import WebFault
from suds.bindings.document import Document
from suds.client import Client
from suds.plugin import MessagePlugin
from suds.sax.element import Element
from suds.sudsobject import asdict
from suds.xsd.sxbase import XBuiltin
from suds.xsd.sxbuiltin import Factory


PATTERN_HEX = r"[0-9a-fA-F]"
PATTERN_ID = r"{hex}{{8}}-{hex}{{4}}-{hex}{{4}}-{hex}{{4}}-{hex}{{12}}".format(
    hex=PATTERN_HEX)
RE_TRANSACTION_ID = compile(PATTERN_ID)


class AuthenticatorPlugin(MessagePlugin):
    """Handle message authentication.

    :param credentials: Stamps API credentials.
    :param wsdl: Configured service client.
    """

    def __init__(self, credentials, client):
        self.credentials = credentials
        self.client = client
        self.authenticator = None

    def marshalled(self, context):
        """Add an authenticator token to the document before it is sent.

        :param context: The current message context.
        """
        body = context.envelope.getChild("Body")
        operation = body[0]

        if operation.name in ("AuthenticateUser", "RegisterAccount"):
            pass
        elif self.authenticator:
            namespace = operation.namespace()
            element = Element("Authenticator", ns=namespace)
            element.setText(self.authenticator)
            operation.insert(element)
        else:
            document = Document(self.client.wsdl)
            method = self.client.service.AuthenticateUser.method
            parameter = document.param_defs(method)[0]
            element = document.mkparam(method, parameter, self.credentials)
            operation.insert(element)

    def unmarshalled(self, context):
        """Store the authenticator token for the next call.

        :param context: The current message context.
        """
        if hasattr(context.reply, "Authenticator"):
            self.authenticator = context.reply.Authenticator
            del context.reply.Authenticator
        else:
            self.authenticator = None

        return context


class BaseService(object):
    """Base service.

    :param configuration: API configuration.
    """

    def __init__(self, configuration):
        Factory.maptag("decimal", XDecimal)
        self.client = Client(configuration.wsdl)
        credentials = self.create("Credentials")
        credentials.IntegrationID = configuration.integration_id
        credentials.Username = configuration.username
        credentials.Password = configuration.password
        self.plugin = AuthenticatorPlugin(credentials, self.client)
        self.client.set_options(plugins=[self.plugin], port=configuration.port)
        self.logger = getLogger("stamps")

    def call(self, method, **kwargs):
        """Call the given web service method.

        :param method: The name of the web service operation to call.
        :param kwargs: Method keyword-argument parameters.
        """
        self.logger.debug("%s(%s)", method, kwargs)
        instance = getattr(self.client.service, method)

        try:
            ret_val = instance(**kwargs)
        except WebFault as error:
            self.logger.warning("Retry %s", method, exc_info=True)
            self.plugin.authenticator = None

            try:  # retry with a re-authenticated user.
                ret_val = instance(**kwargs)
            except WebFault as error:
                self.logger.exception("%s retry failed", method)
                self.plugin.authenticator = None
                raise error

        return ret_val

    def create(self, wsdl_type):
        """Create an object of the given WSDL type.

        :param wsdl_type: The WSDL type to create an object for.
        """
        return self.client.factory.create(wsdl_type)


class StampsService(BaseService):
    """Stamps.com service.
    """

    def add_postage(self, amount, transaction_id=None):
        """Add postage to the account.

        :param amount: The amount of postage to purchase.
        :param transaction_id: Default `None`. ID that may be used to retry the
            purchase of this postage.
        """
        account = self.get_account()
        control = account.AccountInfo.PostageBalance.ControlTotal

        return self.call("PurchasePostage", PurchaseAmount=amount,
                ControlTotal=control, IntegratorTxID=transaction_id)

    def create_add_on(self):
        """Create a new add-on object.
        """
        return self.create("AddOnV7")

    def create_customs(self):
        """Create a new customs object.
        """
        return self.create("CustomsV3")

    def create_array_of_customs_lines(self):
        """Create a new array of customs objects.
        """
        return self.create("ArrayOfCustomsLine")

    def create_customs_lines(self):
        """Create new customs lines.
        """
        return self.create("CustomsLine")

    def create_address(self):
        """Create a new address object.
        """
        return self.create("Address")

    def create_purchase_status(self):
        """Create a new purchase status object.
        """
        return self.create("PurchaseStatus")

    def create_registration(self):
        """Create a new registration object.
        """
        ret_val = self.create("RegisterAccount")
        ret_val.IntegrationID = self.plugin.credentials.IntegrationID
        ret_val.UserName = self.plugin.credentials.Username
        ret_val.Password = self.plugin.credentials.Password

        return ret_val

    def create_shipping(self):
        """Create a new shipping object.
        """
        return self.create("RateV18")

    def get_address(self, address):
        """Get a shipping address.

        :param address: Address instance to get a clean shipping address for.
        """
        return self.call("CleanseAddress", Address=address)

    def get_account(self):
        """Get account information.
        """
        return self.call("GetAccountInfo")

    def get_label(self, from_address, to_address, rate, transaction_id,
            customs=None, sample=False):
        """Get a shipping label.

        :param from_address: The shipping 'from' address.
        :param to_address: The shipping 'to' address.
        :param rate: A rate instance for the shipment.
        :param transaction_id: ID that may be used to retry/rollback the
            purchase of this label.
        :param customs: A customs instance for international shipments.
        :param sample: Default ``False``. Get a sample label without postage.
        """
        return self.call("CreateIndicium", IntegratorTxID=transaction_id,
                Rate=rate, From=from_address, To=to_address, Customs=customs,
                SampleOnly=sample)

    def get_postage_status(self, transaction_id):
        """Get postage purchase status.

        :param transaction_id: The transaction ID returned by
            :meth:`add_postage`.
        """
        return self.call("GetPurchaseStatus", TransactionID=transaction_id)

    def get_rates(self, shipping):
        """Get shipping rates.

        :param shipping: Shipping instance to get rates for.
        """
        rates = self.call("GetRates", Rate=shipping)

        if rates.Rates:
            ret_val = [rate for rate in rates.Rates.Rate]
        else:
            ret_val = []

        return ret_val

    def get_tracking(self, transaction_id):
        """Get tracking events for a shipment.

        :param transaction_id: The transaction ID (or tracking number) returned
            by :meth:`get_label`.
        """
        if RE_TRANSACTION_ID.match(transaction_id):
            arguments = dict(StampsTxID=transaction_id)
        else:
            arguments = dict(TrackingNumber=transaction_id)

        return self.call("TrackShipment", **arguments)

    def register_account(self, registration):
        """Register a new account.

        :param registration: Registration instance.
        """
        arguments = asdict(registration)

        return self.call("RegisterAccount", **arguments)

    def remove_label(self, transaction_id):
        """Cancel a shipping label.

        :param transaction_id: The transaction ID (or tracking number) returned
            by :meth:`get_label`.
        """
        if RE_TRANSACTION_ID.match(transaction_id):
            arguments = dict(StampsTxID=transaction_id)
        else:
            arguments = dict(TrackingNumber=transaction_id)

        return self.call("CancelIndicium", **arguments)


class XDecimal(XBuiltin):
    """Represents an XSD decimal type.
    """

    def translate(self, value, topython=True):
        """Translate between string and decimal values.

        :param value: The value to translate.
        :param topython: Default `True`. Determine whether to translate the
            value for python.
        """
        if topython:
            if isinstance(value, basestring) and len(value):
                ret_val = Decimal(value)
            else:
                ret_val = None
        else:
            if isinstance(value, (int, float, Decimal)):
                ret_val = str(value)
            else:
                ret_val = value

        return ret_val


stamps/config.py
# -*- coding: utf-8 -*-
"""
    stamps.config
    ~~~~~~~~~~~~~

    Stamps.com configuration.

    :copyright: 2014 by Jonathan Zempel.
    :license: BSD, see LICENSE for more details.
"""

from ConfigParser import NoOptionError, NoSectionError, SafeConfigParser
from urllib import pathname2url
from urlparse import urljoin
import os


VERSION = 36


class StampsConfiguration(object):
    """Stamps service configuration. The service configuration may be provided
    directly via parameter values, or it can be read from a configuration file.
    If no parameters are given, the configuration will attempt to read from a
    ``'.stamps.cfg'`` file in the user's HOME directory. Alternately, a
    configuration filename can be passed to the constructor.

    Here is a sample configuration (by default the constructor reads from a
    ``'default'`` section)::

        [default]
        integration_id = XXXXXXXX-1111-2222-3333-YYYYYYYYYYYY
        username = stampy
        password = secret

    :param integration_id: Default `None`. Unique ID, provided by Stamps.com,
        that represents your application.
    :param username: Default `None`. Stamps.com account username.
    :param password: Default `None`. Stamps.com password.
    :param wsdl: Default `None`. WSDL URI. Use ``'testing'`` to use the test
        server WSDL.
    :param port: Default `None`. The name of the WSDL port to use.
    :param file_name: Default `None`. Optional configuration file name.
    :param section: Default ``'default'``. The configuration section to use.
    """

    def __init__(self, integration_id=None, username=None, password=None,
            wsdl=None, port=None, file_name=None, section="default"):
        parser = SafeConfigParser()

        if file_name:
            parser.read([file_name])
        else:
            parser.read([os.path.expanduser("~/.stamps.cfg")])

        self.integration_id = self.__get(parser, section, "integration_id",
                integration_id)
        self.username = self.__get(parser, section, "username", username)
        self.password = self.__get(parser, section, "password", password)
        self.wsdl = self.__get(parser, section, "wsdl", wsdl)
        self.port = self.__get(parser, section, "port", port)

        if self.wsdl is None or wsdl == "testing":
            file_path = os.path.abspath(__file__)
            directory_path = os.path.dirname(file_path)

            if wsdl == "testing":
                file_name = "stamps_v{0}.test.wsdl".format(VERSION)
            else:
                file_name = "stamps_v{0}.wsdl".format(VERSION)

            wsdl = os.path.join(directory_path, "wsdls", file_name)
            self.wsdl = urljoin("file:", pathname2url(wsdl))

        if self.port is None:
            self.port = "SwsimV{0}Soap12".format(VERSION)

        assert self.integration_id
        assert self.username
        assert self.password
        assert self.wsdl
        assert self.port

    @staticmethod
    def __get(parser, section, name, default):
        """Get a configuration value for the named section.

        :param parser: The configuration parser.
        :param section: The section for the given name.
        :param name: The name of the value to retrieve.
        """
        if default:
            vars = {name: default}
        else:
            vars = None

        try:
            ret_val = parser.get(section, name, vars=vars)
        except (NoSectionError, NoOptionError):
            ret_val = default

        return ret_val
